e2e/__tests__/beforeEachAbortsTests.test.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import path from 'path';
+import {wrap} from 'jest-snapshot-serializer-raw';
+import runJest from '../runJest';
+import {extractSummary} from '../Utils';
+
+const dir = path.resolve(__dirname, '../before-each-aborts-tests');
+
+const runAgainstSnapshot = testPath => {
+  const {status, stderr} = runJest(dir, [testPath]);
+  const {summary, rest} = extractSummary(stderr);
+
+  expect(status).toBe(1);
+  expect(wrap(rest)).toMatchSnapshot();
+  expect(wrap(summary)).toMatchSnapshot();
+};
+
+describe('in the global context', () => {
+  test('does not run any tests if a global beforeEach hook fails', () => {
+    runAgainstSnapshot('global/skipsTests.test.js');
+  });
+
+  test('still runs global afterEach hooks if a global beforeEach fails', () => {
+    runAgainstSnapshot('global/afterEachHook.test.js');
+  });
+
+  test('still runs global afterAll hooks if a global beforeEach fails', () => {
+    runAgainstSnapshot('global/afterAllHook.test.js');
+  });
+});
+
+describe('without nested blocks', () => {
+  test('does not run the test if beforeEach fails', () => {
+    runAgainstSnapshot('singleBlock/singleTest.test.js');
+  });
+
+  test('does not run any of the tests if beforeEach fails', () => {
+    runAgainstSnapshot('singleBlock/multipleTests.test.js');
+  });
+
+  test('runs all of the beforeEach hooks if one fails but does not run the tests', () => {
```

Comment:
This test fails in `jest-circus` because it stops running the `beforeEach` hooks once one of them fails.

I didn't fix this because I'm not sure if that's the desired behaviour.

IMO if all `afterEach` and `afterAll` hooks still run I think that all `beforeEach` hooks should run too, otherwise the behaviour is not symmetric. Given that `beforeAll`/`beforeEach` hooks will usually setup tests and `afterAll`/`afterEach` will do clean-up I'd argue that if we're running all the "`teardown`" part we should run all the "`setup`" part.

I'm happy to either change the behaviour in `jest-circus` if you agree with the explanation above or update my PR to make `jest-jasmine2` have the same behaviour as `jest-circus` if you prefer.

Reply:
`beforeEach` and `afterEach` is just sugar for

```
function beforeEach() {
  console.log('before');
}

function afterEach() {
  console.log('after');
}

function test() {
  beforeEach();
  expect(1).toBe(1);
  afterEach();
}
```

(they can be async and Jest waits, but conceptually this is what happens).

I'm not sure we should assign any semantic "setup" and "teardown" meaning to the lifecycle functions. So (IMO) we should just bail if `before` fails and _not_ run `after`.

---

@thymikee @cpojer @scotthovestadt @jeysal @mattphillips do you agree or disagree?

Reply:
I'm not sure if there's a good solution to that. Both have issues. `before` may start some listeners before failing and: 
- they won't be cleared by `after` causing a leak (may surface in watch mode)
- `after` will try to run after failing `before` and fail as well, because the functionality expected to be there is not present

Since Jest already has a mechanism to listen for open connections after test run, I'd lean towards @SimenB solution. So don't treat `after` as a "teardown".

Reply:
@lucasfcosta Just to make sure I’m understanding correctly , you are saying that if **one** `before` hook fails then the rest are ignored but **all** of the `after` hooks are still ran?

If this is the case then I agree, we probably do not want to run any `after` hooks that haven’t had an equivalent`before` hook already ran. 

If we do run the `after` hook without running the equivalent `before` we can end up blowing up because something the `after` depends on isn’t there.

Reply:
I see. I agree with you both. I think that the way @SimenB described it was very interesting to see how the mechanism is actually run but I'd be very careful when translating these concepts into alternative pieces of code as they might become inaccurate. In the excerpt above, for example, the `beforeEach` failing would demonstrate that a `throw` occurred and we're out of the normal execution flow, but if the same happened on an `afterEach` followed by another, there would be a different behaviour.

I think that you are correct especially due to the item `2` mentioned by @thymikee. Even if we treated `after` as a `teardown` it should indeed fail if its setup has not been made or was interrupted somehow.

tl;dr I'm convinced that the adequate behaviour would be as both of you preferred: a failing `beforeEach` should interrupt the next ones in the sequence.

Reply:
I think the major points have already been mentioned. Side note: The `before`, `after` API that we've had even a long time before Jest seems flawed to me because of things like this (and because it does not help structure tests well). The API should look more like React's `useEffect`, associating a teardown operation with its setup by having it as a return value or second parameter.

Reply:
Hi @mattphillips, sorry it seems like there was a bit of a race condition in this discussion 😆 
I ended up posting without seeing your comment since you submitted after I started writing but before I finished 😅

Please see my last comment in this PR, thanks 😊